pkg/kubectl/resource_filter.go

```diff
@@ -0,0 +1,107 @@
+package kubectl
+
+import (
+	"fmt"
+	"io"
+	"reflect"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/v1"
+	"k8s.io/kubernetes/pkg/runtime"
+)
+
+// FilterFunc is a function that knows how to filter a specific resource kind.
+// It receives a generic runtime.Object which must be type-checked by the function.
+// Returns a boolean value true if a resource is filtered, or false otherwise.
+type FilterFunc func(runtime.Object, PrintOptions) bool
+
+// ResourceFilter filters api resources
+type ResourceFilter interface {
+	Filter(obj runtime.Object) (bool, error)
+	AddFilter(reflect.Type, FilterFunc) error
+	PrintFilterCount(io.Writer, string) error
+}
+
+// filterOptions is an implementation of ResourceFilter which contains
+// a map of FilterFuncs for given resource types, PrintOptions for a
+// resource, and a hiddenObjNum to keep track of the number of filtered resources
+type filterOptions struct {
+	filterMap    map[reflect.Type]FilterFunc
```

Comment:
It's so that I can tie the FilterFunc to a resource type, rather than the entire object. Also, if I use `runtime.Object` as the key there would be no way for me to `AddFilter` since I would need the value of the object beforehand